torch/utils/mem_tracker.py

```diff
@@ -0,0 +1,510 @@
+
+
+@dataclass
+class WeakRefInfo:
+    def __init__(
+        self, size: int, element_size: int, device: torch.device, reftype: RefType
+    ) -> None:
+        self.size = size
+        self.element_size = element_size
+        self.reftype = reftype
+        self.device = device
+        self.mem_consumed = self._calculate_mem_consumed()
+
+    def _calculate_mem_consumed(self) -> int:
+        mem = self.size * self.element_size
+        if self.device.type == "cuda":
+            return math.ceil((mem) / _PYTORCH_MIN_ALLOCATE) * _PYTORCH_MIN_ALLOCATE
+        return mem
+
+    def get_mem_consumed(self, st: torch.UntypedStorage) -> int:
+        if st.size() != self.size:
+            self.size = st.size()
+            self.mem_consumed = self._calculate_mem_consumed()
+        return self.mem_consumed
+
+    @staticmethod
+    def get_untyped_storage(t: torch.Tensor) -> torch.UntypedStorage:
+        if isinstance(t, DTensor):
+            t = t._local_tensor
+        return t.untyped_storage()
+
+    @classmethod
+    def get_winfo_if_exists(
+        cls, st: torch.UntypedStorage, WINFO: WeakIdKeyDictionary
+    ) -> Optional[Self]:
+        winfo = WINFO.get(st, None)
+        if winfo is None:
+            return None
+        return winfo
+
+    @classmethod
+    def create_winfo(
+        cls, st: torch.UntypedStorage, reftype: RefType, WINFO: WeakIdKeyDictionary
+    ) -> Self:
+        winfo = cls(st.size(), st.element_size(), st.device, reftype)
+        WINFO[st] = winfo
+        return winfo
+
+    @classmethod
+    def update_and_maybe_create_winfo(
+        cls,
+        t: torch.Tensor,
+        reftype: RefType,
+        WINFO: WeakIdKeyDictionary,
+        update_existing: bool = False,
+    ) -> Self:
+        st = cls.get_untyped_storage(t)
+        if winfo := cls.get_winfo_if_exists(st, WINFO):
+            winfo.reftype = reftype
+            return winfo
+        elif update_existing:
+            raise ValueError("No existing winfo found")
+        else:
+            winfo = cls.create_winfo(st, reftype, WINFO)
+            return winfo
+
+
+def _get_mem_divisor(units: str) -> int:
+    return {"B": 1, "KB": 2**10, "MB": 2**20, "GB": 2**30}.get(units, 1)
+
+
+def _rounding_fn(value: int, divisor: int, precision: int) -> Union[float, int]:
+    return value if divisor == 1 else round(value / divisor, precision)
+
+
+def _print_snapshot(snapshot: Dict[torch.device, Dict[str, int]], units: str):
+    divisor = _get_mem_divisor(units)
+    for dev, dev_snap in snapshot.items():
+        print(
+            f"Device: {dev}",
+            *(
+                f"\t{k}: {_rounding_fn(v, divisor, 2)} {units}"
+                for k, v in dev_snap.items()
+            ),
+            sep="\n",
+        )
+
+
+class MemTracker(TorchDispatchMode):
+    """
+    A TorchDispatchMode that tracks the memory usage of parameters, buffers, activations, gradients, and optimizer states
```

Comment:
We would need to define "parameters, buffers, activations, gradients, and optimizer states" here.
I would say: track all memory usage and tries to associate it to these categories.
With an out-of-line definition of the categories and what is our heuristic to categorize things.

Reply:
> Also you have "temp" in the code that should either be added here or removed.
Will add temp.

Reply:
When you say out-of-line definition where does it exactly go?

Reply:
Where you think it is best as long as you can easily link to it.
It can be in the top level rst file if no docstring is appropriate. Or the docstring of the object defining these.